torchgeo/datasets/geo.py

```diff
@@ -676,46 +780,70 @@ def _load_image(self, index: int) -> Tuple[Tensor, Tensor]:
         return tensor, label
 
 
-class ZipDataset(GeoDataset):
-    """Dataset for merging two or more GeoDatasets.
+class IntersectionDataset(GeoDataset):
+    """Dataset representing the intersection of two GeoDatasets.
+
+    This allows users to do things like:
+
+    * Combine image and target labels and sample from both simultaneously
+      (e.g. Landsat and CDL)
+    * Combine datasets for multiple image sources for multimodal learning or data fusion
+      (e.g. Landsat and Sentinel)
+
+    These combinations require that all queries are present in *both* datasets,
+    and can be combined using an :class:`IntersectionDataset`:
 
-    For example, this allows you to combine an image source like Landsat8 with a target
-    label like CDL.
+    .. code-block:: python
+
+       dataset = landsat & cdl
     """
 
-    def __init__(self, datasets: Sequence[GeoDataset]) -> None:
+    def __init__(
+        self,
+        dataset1: GeoDataset,
+        dataset2: GeoDataset,
+        collate_fn: Callable[
+            [Sequence[Dict[str, Any]]], Dict[str, Any]
+        ] = concat_samples,
+    ) -> None:
         """Initialize a new Dataset instance.
 
         Args:
-            datasets: list of datasets to merge
+            dataset1: the first dataset
+            dataset2: the second dataset
+            collate_fn: function used to collate samples
```

Comment:
Can be a custom function (not just `concat_samples`, `stack_samplers`, or `merge_sampels`) for those weird applications out there.